lib/ocrypto/ec_key_pair.go

```diff
@@ -97,3 +103,122 @@ func (keyPair ECKeyPair) KeySize() (int, error) {
+func ConvertToECDHPublicKey(key interface{}) (*ecdh.PublicKey, error) {
+	switch k := key.(type) {
+	case *ecdsa.PublicKey:
+		// Convert from ecdsa.PublicKey to ECDHPublicKey
+		return k.ECDH()
+	case *ecdh.PublicKey:
+		// No conversion needed
+		return k, nil
+	default:
+		return nil, fmt.Errorf("unsupported public key type")
+	}
+}
+
+// ConvertToECDHPrivateKey convert the ec private key to ECDH private key
+func ConvertToECDHPrivateKey(key interface{}) (*ecdh.PrivateKey, error) {
+	switch k := key.(type) {
+	case *ecdsa.PrivateKey:
+		// Convert from ecdsa.PublicKey to ECDHPublicKey
+		return k.ECDH()
+	case *ecdh.PrivateKey:
+		// No conversion needed
+		return k, nil
+	default:
+		return nil, fmt.Errorf("unsupported private key type")
+	}
+}
+
+// CalculateHKDF generate a key using key derivation function.
+func CalculateHKDF(salt []byte, secret []byte, keyLen int) ([]byte, error) {
+	hkdfObj := hkdf.New(sha256.New, secret, salt, nil)
+
+	derivedKey := make([]byte, keyLen)
+	_, err := io.ReadFull(hkdfObj, derivedKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive hkdf key: %w", err)
+	}
+
+	return derivedKey, nil
+}
+
+// ECPubKeyFromPem generate ec public from pem format
+func ECPubKeyFromPem(pemECPubKey string) (*ecdh.PublicKey, error) {
+	block, _ := pem.Decode([]byte(pemECPubKey))
+	if block == nil {
+		return nil, fmt.Errorf("failed to parse PEM formatted public key")
+	}
+
+	var pub any
+	if strings.Contains(pemECPubKey, "BEGIN CERTIFICATE") {
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("x509.ParseCertificate failed: %w", err)
+		}
+
+		var ok bool
+		if pub, ok = cert.PublicKey.(*ecdsa.PublicKey); !ok {
+			return nil, fmt.Errorf("failed to parse PEM formatted public key")
+		}
+	} else {
+		var err error
+		pub, err = x509.ParsePKIXPublicKey(block.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("x509.ParsePKIXPublicKey failed: %w", err)
+		}
+	}
+
+	switch pub := pub.(type) {
+	case *ecdsa.PublicKey:
+		return ConvertToECDHPublicKey(pub)
+	default:
+		break
+	}
+
+	return nil, fmt.Errorf("not an ec PEM formatted public key")
+}
+
+// ECPrivateKeyFromPem generate ec private from pem format
+func ECPrivateKeyFromPem(privateECKeyInPem string) (*ecdh.PrivateKey, error) {
+	block, _ := pem.Decode([]byte(privateECKeyInPem))
+	if block == nil {
+		return nil, fmt.Errorf("failed to parse PEM formatted private key")
+	}
+
+	priv, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("x509.ParsePKCS8PrivateKey failed: %w", err)
+	}
+
+	switch privateKey := priv.(type) {
+	case *ecdsa.PrivateKey:
+		return ConvertToECDHPrivateKey(privateKey)
+	default:
+		break
+	}
+
+	return nil, fmt.Errorf("not an ec PEM formatted private key")
+}
+
+// ComputeECDHKey calculate shared secret from public key from one party and the private key from another party.
+func ComputeECDHKey(privateKeyInPem string, publicKeyInPem string) ([]byte, error) {
```

Comment:
I have questions around this as well. I think we need to be careful wrapping functionality in to many helpers as this can sometimes make things more difficult to follow and understand.  

What is the use case for this function?

Reply:
The use case: SDK public key(pem) from rewrap request with KAS private key to generate symmetric key. Both are in pem formats.